Point_set_processing_3/include/CGAL/estimate_lfs.h

```diff
@@ -0,0 +1,687 @@
+    std::vector<Point> intersections;
+    bool flag = recursive_dichotomic_search(seg, Point_2(0.0, ls), Point_2(l, lt), neighbor_query,
+                                    epsilon_band, intersections, lipschitz);
+
+    if (flag && intersections.size() >= 2)
+    {
+      std::vector<FT> dsqs;
+      for (std::size_t i = 0; i < intersections.size(); i++)
+      {
+        const auto intersection = intersections[i];
+
+        FT dsq = (query - intersection).squared_length();
+        dsqs.push_back(dsq);
+      }
+      std::sort(dsqs.begin(), dsqs.end(), std::less<FT>());
+      // dist should be larger than epsilon_band
+      auto it = upper_bound(dsqs.begin(), dsqs.end(), project_radius*project_radius);
+
+      if (it != dsqs.end())
+      {
+        FT dsq_min = *it;
+        antipodal_point_dsqs.push_back(dsq_min);
+      }
+    }
+  }
+
+  FT dsq_upper_bound = 4.0 * bsphere.squared_radius();
+  
+  if(antipodal_point_dsqs.empty())
+        antipodal_point_dsqs.push_back(dsq_upper_bound);
+
+  // robust distance function here
+  FT sum_squared_distances = std::accumulate(antipodal_point_dsqs.begin(), antipodal_point_dsqs.end(), 0.0);
+  std::size_t sz = antipodal_point_dsqs.size();
+  if (sz == 0)
+      return 0.0;
+  else
+      return std::sqrt(sum_squared_distances / sz);
+}
+
+template <typename Monge_form>
+typename Monge_form::FT
+min_curvature_radius(Monge_form &monge_form)
+{
+  // basic geometric types
+  typedef typename Monge_form::FT FT;
+    
+  FT max_principal_curvature = monge_form.principal_curvatures(0);
+  FT min_principal_curvature = monge_form.principal_curvatures(1);
+
+  FT r1 = 1.0 / std::abs(max_principal_curvature);
+  FT r2 = 1.0 / std::abs(min_principal_curvature);
+  FT r = (std::min)(r1, r2);
+
+  return r;
+}
+
+template <typename NeighborQuery>
+typename NeighborQuery::FT
+estimate_local_feature_size(const typename NeighborQuery::Point_3& query, ///< point 
+                    typename NeighborQuery::Kernel::Vector_3& normal,
+                    const NeighborQuery& neighbor_query, ///< KD-tree
+                    const typename NeighborQuery::Kernel::Sphere_3& bsphere, ///< bounding sphere
+                    unsigned int jet_k, ///< number of neighbors
+                    typename NeighborQuery::FT neighbor_radius,
+                    unsigned int degree_fitting,
+                    unsigned int degree_monge,
+                    typename NeighborQuery::FT epsilon_band,
+                    typename NeighborQuery::FT apex_angle,
+                    std::size_t N_rays
+                    )
+{
+    // basic geometric types
+    typedef typename NeighborQuery::Kernel Kernel;
+    typedef typename Kernel::Point_3  Point;
+    typedef typename Kernel::Vector_3 Vector;
+    typedef typename Kernel::FT FT;
+
+    // types for jet fitting
+    typedef Monge_via_jet_fitting< Kernel> Monge_jet_fitting;
+    typedef typename Monge_jet_fitting::Monge_form Monge_form;
+
+    std::vector<Point> points;
```

Comment:
Maybe call `reserve()`

Reply:
if using sphere search, the size of the vector is not known.